lib/src/mock_query.dart

```diff
@@ -1,45 +1,63 @@
 import 'dart:math';
 
 import 'package:cloud_firestore/cloud_firestore.dart';
-import 'package:cloud_firestore_mocks/cloud_firestore_mocks.dart';
 import 'package:cloud_firestore_platform_interface/cloud_firestore_platform_interface.dart';
 import 'package:mockito/mockito.dart';
 
 import 'mock_snapshot.dart';
 
+typedef List<DocumentSnapshot> _QueryOperation(List<DocumentSnapshot> input);
+
 class MockQuery extends Mock implements Query {
-  List<DocumentSnapshot> documents;
-  final MockFirestoreInstance _firestore;
+  /// Previous query in a Firestore query chain. Null if this instance is a
+  /// collection reference. A query chain always starts with a collection
+  /// reference, which does not have a previous query.
+  final Query _parentQuery;
+
+  /// Operation to perform in this query, such as "where", "limit", and
+  /// "orderBy". Null if this is a collection reference.
+  final _QueryOperation _operation;
 
-  MockQuery(this._firestore, this.documents);
+  MockQuery([this._parentQuery, this._operation]);
 
   // ignore: unused_field
   final QueryPlatform _delegate = null;
 
   @override
-  Future<QuerySnapshot> getDocuments({Source source = Source.serverAndCache}) {
-    final savedDocuments = documents.where((snapshot) => _firestore.hasSavedDocument(snapshot.reference.path)).toList();
```

Comment:
Now Query is not responsible to check whether a document is saved or not. It is collection references' responsibility to check that, and a collection reference is always at the beginning of a query chain.